StackOverFlow in JSP Caused by Recurisve JspContextWrapper#include call

**Describe the bug**  
```
SRVE0777E: Exception thrown by application class 'org.apache.jasper.runtime.PageContextImpl.handlePageException:730'
com.ibm.websphere.servlet.error.ServletErrorReport: java.lang.StackOverflowError
	at org.apache.jasper.runtime.PageContextImpl.handlePageException(PageContextImpl.java:730)
        ...
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.lang.Thread.run(Thread.java:825)
Caused by: java.lang.StackOverflowError
	at org.apache.jasper.runtime.JspContextWrapper.include(JspContextWrapper.java:327)
	at org.apache.jasper.runtime.JspContextWrapper.include(JspContextWrapper.java:327)
	at org.apache.jasper.runtime.JspContextWrapper.include(JspContextWrapper.java:327)
```

**Steps to Reproduce**  
A JSP tag calls the following method:
[PageContext#include(String relativeUrlPath, boolean flush)](https://javaee.github.io/javaee-spec/javadocs/javax/servlet/jsp/PageContext.html#include-java.lang.String-boolean-)

**Expected behavior**  
A StackOverflowError should not occur. 

**Diagnostic information:**  
 - OpenLiberty Version: All 
 - Affected feature(s) jsp-2.3, pages-3.0, pages-3.1
 - Java Version: N/A
 - server.xml configuration: N/A
 
**Additional context**  
The code should the wrapped page context object instead. 
Also see the original BZ Issue:`https://bz.apache.org/bugzilla/show_bug.cgi?id=45666` (over 15 years old!)

--- a/dev/com.ibm.ws.jsp.2.3/src/org/apache/jasper/runtime/JspContextWrapper.java
+++ b/dev/com.ibm.ws.jsp.2.3/src/org/apache/jasper/runtime/JspContextWrapper.java
@@ -324,7 +324,7 @@ public class JspContextWrapper extends PageContext implements VariableResolver {
 
     public void include(String relativeUrlPath, boolean flush) 
 	    throws ServletException, IOException {
-	include(relativeUrlPath, false); // XXX
+	invokingJspCtxt.include(relativeUrlPath, false);
     }
 
     public VariableResolver getVariableResolver() {

--- a/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java
+++ b/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTests.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2013, 2023 IBM Corporation and others.
+ * Copyright (c) 2013, 2024 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -66,6 +66,7 @@ public class JSPTests {
     private static final String TestJDT_APP_NAME = "TestJDT";
     private static final String OLGH20509_APP_NAME1 = "OLGH20509jar";
     private static final String OLGH20509_APP_NAME2 = "OLGH20509TDfalse";
+    private static final String OLGH27779_APP_NAME = "OLGH27779";
 
     @Server("jspServer")
     public static LibertyServer server;
@@ -89,6 +90,8 @@ public class JSPTests {
         ShrinkHelper.defaultDropinApp(server, TestJDT_APP_NAME + ".war");
 
         ShrinkHelper.defaultDropinApp(server, TestEDR_APP_NAME + ".war");
+
+        ShrinkHelper.defaultDropinApp(server, OLGH27779_APP_NAME + ".war");
 
         JavaArchive jspJar = ShrinkWrap.create(JavaArchive.class, "OLGH20509Include.jar");
         jspJar = (JavaArchive) ShrinkHelper.addDirectory(jspJar, "test-applications/includejar/resources");
@@ -985,6 +988,20 @@ public class JSPTests {
         server.deleteFileFromLibertyServerRoot(relEdrPath + orgEdrFile); // cleanup
         Thread.sleep(500L); // ensure file is deleted
     }
+
+    /*
+     * Verify a stackover flow error does not occur via the include(String relativeUrlPath, boolean flush) 
+     * 
+     * See https://github.com/OpenLiberty/open-liberty/issues/27779 for more details
+     */
+    @Test
+    @Mode(TestMode.FULL)
+    public void testOLGH27779() throws Exception {
+        this.verifyStringInResponse(OLGH27779_APP_NAME, "index.jsp", "Test Passed!");
+    }
+
+
+    // Helper Methods
 
     public void makeConcurrentRequests(WebConversation wc1, WebRequest request1, int numberOfCalls) throws Exception {
         final ExecutorService executor = Executors.newFixedThreadPool(numberOfCalls);

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/WEB-INF/tags/include.tag
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/WEB-INF/tags/include.tag
@@ -1,0 +1,13 @@
+<!--
+    Copyright (c) 2024 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<%
+// Can only be invoked from a tag 
+((PageContext) jspContext).include("/include.jsp", false);
+%>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/WEB-INF/web.xml
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/WEB-INF/web.xml
@@ -1,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2023 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<web-app
+    version="3.1"
+    xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd">
+
+  <display-name>OLGH26891</display-name>
+
+</web-app>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/include.jsp
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/include.jsp
@@ -1,0 +1,10 @@
+<!--
+    Copyright (c) 2024 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+Test Passed!

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/index.jsp
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/OLGH27779.war/resources/index.jsp
@@ -1,0 +1,21 @@
+<!--
+    Copyright (c) 2024 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+ <%@ page language="java" contentType="text/html; charset=ISO-8859-1"
+ pageEncoding="ISO-8859-1"%>
+ <%@taglib prefix="tag" tagdir="/WEB-INF/tags" %>
+<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
+<html>
+<head>
+<title> Test JspContextWrapper#Include </title>
+</head>
+<body>
+   <tag:include/>
+</body>
+</html>

--- a/dev/com.ibm.ws.jsp/src/org/apache/jasper/runtime/JspContextWrapper.java
+++ b/dev/com.ibm.ws.jsp/src/org/apache/jasper/runtime/JspContextWrapper.java
@@ -324,7 +324,7 @@ public class JspContextWrapper extends PageContext implements VariableResolver {
 
     public void include(String relativeUrlPath, boolean flush) 
 	    throws ServletException, IOException {
-	include(relativeUrlPath, false); // XXX
+	invokingJspCtxt.include(relativeUrlPath, false);
     }
 
     public VariableResolver getVariableResolver() {

--- a/dev/io.openliberty.pages.3.1.internal.factories/src/org/apache/jasper/runtime/JspContextWrapper.java
+++ b/dev/io.openliberty.pages.3.1.internal.factories/src/org/apache/jasper/runtime/JspContextWrapper.java
@@ -327,7 +327,7 @@ public class JspContextWrapper extends PageContext implements VariableResolver {
 
     public void include(String relativeUrlPath, boolean flush)
 	    throws ServletException, IOException {
-	include(relativeUrlPath, false); // XXX
+	invokingJspCtxt.include(relativeUrlPath, false);
     }
 
     public VariableResolver getVariableResolver() {

